### PR TITLE
SQL, IB: Copy history via CREATE .. SELECT [closes #157, #152]

### LINKED ISSUE #157: Versioned CREATE TABLE ... SELECT
This should create versioned table and copy history correctly if SELECT contains FOR SYSTEM_TIME clause.

- [x] pass system fields to CREATE;
- [x] modes of `vers_hide`: AUTO, IMPLICIT, FULL, NEVER;
- [x] InnoDB engine;
- [x] selecting system fields from different tables must be prohibited;

### PR DESCRIPTION
**Pull request by midenok**
Please, review.

### CODE REVIEW

File: sql/mysqld.h [hunk 1]
```diff
@@ -176,7 +176,13 @@ extern const char *log_output_str;
 extern const char *log_backup_output_str;
 extern char *temporal_current_timestamp;
 extern my_bool vers_force;
-extern my_bool vers_hide;
+enum vers_hide_enum: ulong {
```
Comment (kevgs) [inline sql/mysqld.h]: Specifying a type of enum is a c++11 feature.

Comment (midenok) [reply sql/mysqld.h]: I know. Could you propose different solution?

File: plugin/versioning/versioning.cc [hunk 2]
```diff
@@ -30,15 +30,15 @@ static int forced_versioning_init(void *p __attribute__ ((unused)))
 
   DBUG_ENTER("forced_versioning_init");
   vers_force= true;
```
Comment (kevgs) [inline plugin/versioning/versioning.cc:32]: It's possible to load a plugin through [queries](https://mariadb.com/kb/en/mariadb/plugin-commands/)
And it looks like a data race when someone `CREATE TABLE` while plugin is loading.

File: sql/mysqld.h (hunk 1 above)
Comment (kevgs) [reply sql/mysqld.h]:
```diff
diff --git a/sql/mysqld.cc b/sql/mysqld.cc
index e48988a..ce68980 100644
--- a/sql/mysqld.cc
+++ b/sql/mysqld.cc
@@ -786,7 +786,7 @@ char *opt_logname, *opt_slow_logname, *opt_bin_logname;
 /* System Versioning */
 char *temporal_current_timestamp;
 my_bool vers_force= false;
-vers_hide_enum vers_hide= VERS_HIDE_AUTO;
+ulong vers_hide= VERS_HIDE_AUTO;
 
 /* Static variables */
 
diff --git a/sql/mysqld.h b/sql/mysqld.h
index e06b9fa..0e51203 100644
--- a/sql/mysqld.h
+++ b/sql/mysqld.h
@@ -176,13 +176,11 @@ extern const char *log_output_str;
 extern const char *log_backup_output_str;
 extern char *temporal_current_timestamp;
 extern my_bool vers_force;
-enum vers_hide_enum: ulong {
-  VERS_HIDE_AUTO= 0,
-  VERS_HIDE_IMPLICIT,
-  VERS_HIDE_FULL,
-  VERS_HIDE_NEVER
-};
-extern vers_hide_enum vers_hide;
+const static ulong VERS_HIDE_AUTO= 0;
+const static ulong VERS_HIDE_IMPLICIT= 1;
+const static ulong VERS_HIDE_FULL= 2;
+const static ulong VERS_HIDE_NEVER= 3;
+extern ulong vers_hide;
 extern char *mysql_home_ptr, *pidfile_name_ptr;
 extern MYSQL_PLUGIN_IMPORT char glob_hostname[FN_REFLEN];
 extern char mysql_home[FN_REFLEN]
```

File: sql/sys_vars.cc [hunk 3]
```diff
@@ -391,9 +391,14 @@ static Sys_var_mybool Sys_vers_force(
        "vers_force", "Force system versioning for all created tables",
        GLOBAL_VAR(vers_force), CMD_LINE(OPT_ARG), DEFAULT(FALSE));
 
-static Sys_var_mybool Sys_vers_hide(
-       "vers_hide", "Hide system versioning from being displayed in table info",
-       GLOBAL_VAR(vers_hide), CMD_LINE(OPT_ARG), DEFAULT(FALSE));
+static const char *vers_hide_keywords[]= {"AUTO", "IMPLICIT", "FULL", "NEVER", NullS};
```
Comment (kevgs) [inline sql/sys_vars.cc:394]: I guess `NullS` presents here for consistency. But why being consistent with a strange ancient MySQL stuff? Everybody knows about `NULL`. Using `NullS` makes code harder to read.

File: plugin/versioning/versioning.cc (hunk 2 above)
Comment (midenok) [reply plugin/versioning/versioning.cc:32]: No, there is no race. Besides, there is no difference from simple `set global vers_force = 1;`.

File: sql/sys_vars.cc (hunk 3 above)
Comment (midenok) [reply sql/sys_vars.cc:394]: I don't see any trouble with it. It's self-explanatory.

File: sql/mysqld.h (hunk 1 above)
Comment (midenok) [reply sql/mysqld.h]: Why it works with other enums (f.ex. `enum_binlog_format`, see `Sys_binlog_format`)? We need to convert our sysvars to `SESSION_VAR` in future anyway.

Comment (kevgs) [reply sql/mysqld.h]: It do not work with `enum_binlog_format` and actually uses a different `ulong` [variable](https://github.com/tempesta-tech/mariadb_10.2/blob/427f834841f48cb0185a23536a6a59e8fb7cd221/sql/sql_class.h#L698).

This is even better.
```diff
diff --git a/sql/mysqld.cc b/sql/mysqld.cc
index e48988a..ce68980 100644
--- a/sql/mysqld.cc
+++ b/sql/mysqld.cc
@@ -786,7 +786,7 @@ char *opt_logname, *opt_slow_logname, *opt_bin_logname;
 /* System Versioning */
 char *temporal_current_timestamp;
 my_bool vers_force= false;
-vers_hide_enum vers_hide= VERS_HIDE_AUTO;
+ulong vers_hide= VERS_HIDE_AUTO;
 
 /* Static variables */
 
diff --git a/sql/mysqld.h b/sql/mysqld.h
index e06b9fa..2230ee2 100644
--- a/sql/mysqld.h
+++ b/sql/mysqld.h
@@ -176,13 +176,13 @@ extern const char *log_output_str;
 extern const char *log_backup_output_str;
 extern char *temporal_current_timestamp;
 extern my_bool vers_force;
-enum vers_hide_enum: ulong {
+enum vers_hide_enum {
   VERS_HIDE_AUTO= 0,
   VERS_HIDE_IMPLICIT,
   VERS_HIDE_FULL,
   VERS_HIDE_NEVER
 };
-extern vers_hide_enum vers_hide;
+extern ulong vers_hide;
 extern char *mysql_home_ptr, *pidfile_name_ptr;
 extern MYSQL_PLUGIN_IMPORT char glob_hostname[FN_REFLEN];
 extern char mysql_home[FN_REFLEN];
```

Comment (midenok) [reply sql/mysqld.h]: Fixed.

Comment (midenok) [reply sql/mysqld.h]: Btw, `Sys_binlog_format` demonstrates how to detect explicit setting (via `ON_UPDATE` hook).

**Review by kevgs — APPROVED** [review]

**Pull request merged**